### PR TITLE
Improve Enums performance

### PR DESCRIPTION
**Pull request by fpagnoux**
Connected to openfisca/openfisca-france#831

Improve performance of changes introduced in `21.0`.

For changelog clarity, I recommand unpublishing the intertermediate step `21.0`, as no one is using it.

### CODE REVIEW

**Comment by fpagnoux** [comment]
@sandcha do you have time to look at this to move forward with the enums ? Thanks !

File: openfisca_core/indexed_enums.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    __gt__ = _forbidden_operation
+    __ge__ = _forbidden_operation
+    __and__ = _forbidden_operation
+    __or__ = _forbidden_operation
+
+    def decode(self):
+        """
+            Return the array of enum items corresponding to self
+
+            >>> enum_array
```
Comment (Anna-Livia) [inline openfisca_core/indexed_enums.py]: Shouldn't it be : 
```
>>> enum_array = [<HousingOccupancyStatus.free_lodger: u'Free logder'>]
>>> EnumArray(enum_array)
>>> enum_array[0]
....
```
?

**Comment by Anna-Livia** [comment]
Also, could you share the performance measures and how you obtained them?
That way we could re-use them :)

File: CHANGELOG.md [hunk 2]
```diff
@@ -61,37 +66,36 @@ simulation = tax_benefit_system.new_scenario().init_single_entity(
 > ```
 >
 
-#### Before and after
+##### Change the simulation inputs and outputs for enumeration variables
```
[the remark this answers is not in the document]
Comment (fpagnoux) [reply CHANGELOG.md]: Yep, both the input and the output format have changed compared to the `v20`. (For instance for the Web API, you'd get back the label `"Free lodger"`, now you get the identifier `"free_lodger"`

I added "Change" to uniformise the two subtitles: 
- _Change the way enumerations (Enum) are defined when coding variables_
- _Change the simulation inputs and outputs for enumeration variables_.

File: openfisca_core/indexed_enums.py (hunk 1 above)
Comment (fpagnoux) [reply openfisca_core/indexed_enums.py]: Right, a first line to explain where this `enum_array` is coming from could help 👍

Comment (fpagnoux) [reply openfisca_core/indexed_enums.py]: Example simplified

File: openfisca_core/indexed_enums.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    # Tweak enums to add an index attribute to each enum item
+    def __init__(self, name):
+        # When the enum item is initialized, self._member_names_ contains the names of the previously initialized items, so its length is the index of this item.
+        self.index = len(self._member_names_)
+
+    # Bypass the slow Enum.__eq__
+    __eq__ = object.__eq__
+
+    @classmethod
+    def encode(cls, array):
```
[the remark this answers is not in the document]
Comment (fpagnoux) [reply openfisca_core/indexed_enums.py]: Done !

**Comment by fpagnoux** [comment]
#### First script showing that enum comparison is much slower than int comparison

```py
from openfisca_france import CountryTaxBenefitSystem

import numpy as np
import time

N = 5000
tbs = CountryTaxBenefitSystem()

categorie_salarie_int = np.floor(np.random.rand(4 * N) * 8)
enum = tbs.get_variable('categorie_salarie').possible_values
categorie_salarie_enum = np.select([categorie_salarie_int == i for i in range(8)], enum)


def test_elem():
	def formula_before(categorie_salarie):
		return np.select(
			[categorie_salarie == i for i in range(8)],
			range(100, 900, 100)
			)

	def formula_after(categorie_salarie):
		return np.select(
			[categorie_salarie == i for i in enum],
			range(100, 900, 100)
			)

	start_time = time.time()
	formula_before(categorie_salarie_int)
	t1 = time.time() - start_time
	print("BEFORE: --- %s seconds ---" % (t1))


	start_time_2 = time.time()
	formula_after(categorie_salarie_enum)
	t2 = time.time() - start_time_2
	print("AFTER: --- %s seconds ---" % (t2))

	print(t2/t1)

test_elem()
``
`

Typical outcome:
```
BEFORE: --- 0.00224184989929 seconds ---
AFTER: --- 0.177685022354 seconds ---
79.2582154632
```

**Comment by fpagnoux** [comment]
### Encoding to the rescue (still at the operation level)

```py
from openfisca_france import CountryTaxBenefitSystem
import numpy as np
import time

N = 5000
tbs = CountryTaxBenefitSystem()

categorie_salarie_int = np.floor(np.random.rand(4 * N) * 8)
enum = tbs.get_variable('categorie_salarie').possible_values
categorie_salarie_enum = np.select([categorie_salarie_int == i for i in range(8)], enum)
categorie_salarie_enum = enum.encode(categorie_salarie_enum)


def test_elem():
	def formula_before(categorie_salarie):
		return np.select(
			[categorie_salarie == i for i in range(8)],
			range(100, 900, 100)
			)

	def formula_after(categorie_salarie):
		return np.select(
			[categorie_salarie == i for i in enum],
			range(100, 900, 100)
			)

	start_time = time.time()
	formula_before(categorie_salarie_int)
	t1 = time.time() - start_time
	print("BEFORE: --- %s seconds ---" % (t1))


	start_time_2 = time.time()
	formula_after(categorie_salarie_enum)
	t2 = time.time() - start_time_2
	print("AFTER: --- %s seconds ---" % (t2))

	print(t2/t1)

test_elem()
```

Typical outcome:

```
BEFORE: --- 0.00061297416687 seconds ---
AFTER: --- 0.000654220581055 seconds ---
1.06728899261
```

Conclusion : same order of magnitude, with only a slight degradation.

**Comment by fpagnoux** [comment]
### Real condition tests

_Needs https://github.com/openfisca/openfisca-core/pull/604, https://github.com/openfisca/openfisca-core/pull/598, and to use the right `openfisca-france` branch_

```py
from simulation_generator import make_simulation, randomly_init_variable
from openfisca_france import CountryTaxBenefitSystem
from openfisca_core.periods import period
import numpy as np
import time

N = 5000
USE_ENUMS = False  # Swith to True to use enums

tbs = CountryTaxBenefitSystem()

simulation = make_simulation(tbs, 4 * N, N)
randomly_init_variable(simulation, 'salaire_net', 2017, max_value = 50000, condition = simulation.persons.has_role(simulation.famille.DEMANDEUR))

categorie_salarie_int = np.floor(np.random.rand(4 * N) * 8)

if USE_ENUMS:
	enum = tbs.get_variable('categorie_salarie').possible_values
	categorie_salarie_enum = np.select([categorie_salarie_int == i for i in range(8)], enum)
	simulation.persons.get_holder('categorie_salarie').set_input(period(2017), categorie_salarie_enum)
else:
	simulation.persons.get_holder('categorie_salarie').set_input(period(2017), categorie_salarie_int)

def test_aggregate():
	start_time = time.time()
	simulation.calculate('revenu_disponible', "2017")
	print("AFTER: --- %s seconds ---" % (time.time() - start_time))

test_aggregate()

```

**Pull request merged**